### ISSUE TITLE
Redact Screenshots via View Hierarchy

### ISSUE THREAD

**Issue by stefanosiano** [getsentry/sentry-dart#1956]
We have Screenshots, and we have View Hierarchy. For Mobile SR, we use VH to redact SS-based Replays - we should "backport" this functionality to the SS feature.

Rationale: SS can contain PII and other sensitive information, therefore the feature is not viable for some use cases and orgs. Redacting SS via VH enables wider adoption and improves experience for those who already use it.

Let's evaluate:
1) Implement screeshot redacting and use it for session replay later
2) Wait for session replay and reuse its implementation for SS and VH

**Comment by stefanosiano** [comment]
Let's wait for SR plans and prioritize this

**Comment by The-RootCause** [comment]
Any update on this feature. It's very important

**Comment by kahest** [comment]
Hi @The-RootCause thanks for your interest - it's very likely we will implement this in the next few months, we'll keep this issue updated with any progress

**Comment by The-RootCause** [comment]
Amazing, thank @kahest

**Comment by JulianBissekkou** [comment]
I would love to help contribute this feature since we also need this in our app.

**Comment by vaind** [comment]
This was brought to my attention by @buenaflor yesterday...

We already have screenshot redacting implemented for replay. As soon as #2208 is merged, the code can be adapted/reused for  standard screenshots too.

See [WidgetFilter](https://github.com/getsentry/sentry-dart/pull/2208/files#diff-a489f467a5ea2af3ccf5dcfb86c094c37797139480a292d01110b7f52a775253) which walks the tree and creates a list of bounding rectangles to redact. Then it's just the question of [painting them onto the screenshot](https://github.com/getsentry/sentry-dart/pull/2208/files#diff-9ea820a1416f6c2ffd0cae27eb64e26858432e88e2aff494b02ed7739aba0f75R113)

**Comment by buenaflor** [comment]
@vaind do you know the rough timeline for SR? I was thinking if it makes sense to use the `WidgetFilter` implementation now outside of the SR branch so we can already implement it for screenshots

**Comment by vaind** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/6349682?s=64&amp;u=6e3636d7ce3521a5d9ffe6fe96b0a3ad82a2545c&amp;v=4">@vaind](https://github.com/vaind) do you know the rough timeline for SR? I was thinking if it makes sense to use the `WidgetFilter` implementation now outside of the SR branch so we can already implement it for screenshots

I want to make another pre-released today, now that sentry-java has shipped with an update we need. If all goes well, we can merge replay branch to main & still keep it as an alpha feature (i.e. subject to change), since there doesn't seem to be any immediate issue that would require us to roll back these changes. That was the only reason for keeping another branch alive.

**Comment by vaind** [comment]
> See [WidgetFilter](https://github.com/getsentry/sentry-dart/pull/2208/files#diff-a489f467a5ea2af3ccf5dcfb86c094c37797139480a292d01110b7f52a775253) which walks the tree and creates a list of bounding rectangles to redact. Then it's just the question of [painting them onto the screenshot](https://github.com/getsentry/sentry-dart/pull/2208/files#diff-9ea820a1416f6c2ffd0cae27eb64e26858432e88e2aff494b02ed7739aba0f75R113)

This is now merged to main so this issue is unblocked

**Comment by vaind** [comment]
@martinhaintz I've noticed you've assigned this to yourself. Just be aware there will be changes in the replay redaction logic (thus also in screenshot reduction) in order to support user-configurable widget-level redaction

**Comment by buenaflor** [comment]
@vaind should we wait here until you've done the changes?

**Comment by vaind** [comment]
Not sure how far along @martinhaintz is. If he hasn't started yet then yeah, I think it makes sense to deprioritize this for now.
If a lot of work is already done on screenshot redacting (I haven't found a PR though) then maybe it makes sense to finish and merge that 
and I'll fit it together with the changes I'm preparing

**Comment by buenaflor** [comment]
@martinhaintz wdyt?

**Comment by martinhaintz** [comment]
@vaind Sorry, for interrupting your plan. As @buenaflor was on vacation, I spoke with @kahest and he assigned it to me via chat.

I currently have no working code and therefore have not created a branch yet. I haven't invested too much time yet. I started with the issue today.

So I will stop working on it since you are already kind of working on it.

**Comment by vaind** [comment]
Alright, I'll ping you. Sorry for the inconvenience
